### PR TITLE
Feature/2 database design setting up backend

### PR DESCRIPTION
**Pull request by abroa01**
Merging code from feature branch to main with all the backends changes updated

### CODE REVIEW

File: backend/config/database.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+const { MongoClient } = require('mongodb');
+
+let dbConnection;
+const dbUrl = process.env.DATABASE_URL
+module.exports = {
```
Comment (anishapant21) [inline backend/config/database.js:5]: Lets leave an empty space above this line for better readability.

File: backend/controllers/authController.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
+const User = require('../models/user');
+const bcrypt = require('bcrypt');
+const jwt = require('jsonwebtoken');
+const { getDb } = require('../config/database')
+
+
```
Comment (anishapant21) [inline backend/controllers/authController.js:6]: Lets remove this empty space.

File: backend/controllers/authController.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+// User registration
+exports.signup = async (req, res) => {
+  try {
+    const { name, email, password } = req.body;
+
+    const db = getDb()
+    // console.log(db, 'hit')
```
Comment (anishapant21) [inline backend/controllers/authController.js:13]: Lets remove this comment if it's not required.

File: backend/controllers/authController.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    // Check if the user already exists
+    const userExists = await db.collection("user_signup").findOne({ email });
+    console.log(userExists)
```
Comment (anishapant21) [inline backend/controllers/authController.js:16]: Lets remove this console log from here.

File: backend/controllers/authController.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      return res.status(400).json({ error: 'User already exists' });
+    } else {
+      const hashedPassword = await bcrypt.hash(password, 10);
+      await db.collection("user_signup").insertOne({ name, email, hashedPassword })
+    }
+
+    const user = await db.collection("user_signup").findOne({ email });
+
+
+
```
Comment (anishapant21) [inline backend/controllers/authController.js:27]: Lets remove this empty spaces from here.

File: backend/controllers/usersController.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  try {
+    const db = getDb();
+    const user = await db.collection("user_signup").findOne({ _id: new ObjectId(userId) });
+
+    if (!user) {
+      return res.status(404).json({ error: 'User not found' });
+    }
+    res.status(200).json(user);
+  } catch (error) {
+    console.error(error);
```
Comment (anishapant21) [inline backend/controllers/usersController.js:20]: Same as above.

File: backend/controllers/usersController.js [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  const userId = req.params.userId;
+
+  try {
+    const db = getDb();
+    const user = await db.collection("user_signup").findOne({ _id: new ObjectId(userId) });
+
+    if (!user) {
+      return res.status(404).json({ error: 'User not found' });
+    }
+    res.status(200).json(user);
```
Comment (anishapant21) [inline backend/controllers/usersController.js:18]: Lets leave an empty space above this line.

File: backend/controllers/usersController.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+    const passwordMatch = await bcrypt.compare(password, user.password);
+
+    if (!passwordMatch) {
+      return res.status(401).json({ error: 'Invalid email or password' });
+    }
+
+    res.status(200).json(user);
+  } catch (error) {
+    console.error(error);
```
Comment (anishapant21) [inline backend/controllers/usersController.js:65]: Same as above.

File: backend/controllers/usersController.js [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+  try {
+    const db = getDb();
+    const user = await db.collection("user_signup").findOne({ email });
+
+    if (!user) {
+      return res.status(401).json({ error: 'Invalid email or password' });
+    }
+
+    const passwordMatch = await bcrypt.compare(password, user.password);
```
Comment (anishapant21) [inline backend/controllers/usersController.js:57]: Lets move all the assignments at the top before the if condition.

File: backend/models/user.js [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+  password: {
+    type: String,
+    required: true,
+  },
+//   passwordConfirmation: {
+//     type: String,
+//     required: true,
+//   },
+  // Add other fields as needed (e.g., tokens, profile, etc.)
+});
```
Comment (anishapant21) [inline backend/models/user.js:24]: Same as above.

File: backend/models/workout.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    type: String,
+    required: true,
+  },
+  exercises: [
+    {
+      name: String,
+      duration: Number,
+    },
+  ],
+  // Add more fields as needed
```
Comment (anishapant21) [inline backend/models/workout.js:14]: Same here.

File: backend/routes/gymrat.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
+const express = require('express')
+const router = express.Router()
+
+
+//Fetching Basic Details all route
+router.get('/', (req, res) => {
+    res.send('Hello World')
```
Comment (anishapant21) [inline backend/routes/gymrat.js:7]: Same here.

File: backend/routes/gymrat.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+})
+
+//Fetching Basic Details byId route
+router.get('/:id', (req, res) => {
+    // req.params.id
```
Comment (anishapant21) [inline backend/routes/gymrat.js:13]: Same here.

File: backend/routes/gymrat.js [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+router.delete('/:id', (req, res) => {
+    
+})
+
+
+
+
+
+
+module.exports = router
```
Comment (anishapant21) [inline backend/routes/gymrat.js:38]: Same here.

File: backend/routes/users.js [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+const usersController = require('../controllers/usersController');
+
+// Get user profile
+router.get('/:userId', usersController.getUserProfile);
+
+// Update user information
+router.put('/:userId', usersController.updateUser);
+
+// Create a new user
+// router.post('/add', usersController.createUser);
```
Comment (anishapant21) [inline backend/routes/users.js:12]: .

File: backend/server.js [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+connectToDb(dbUrl, (err) => {
+  if (err) {
+    console.error('Error connecting to the database:', err);
+  } else {
+    // Set up routes and start the server
+    app.use('/auth', authRoutes);
+    app.use('/workouts', workoutsRoutes);
+    app.use('/users', usersRoutes);
+    app.use('/gymrat', gymRatRouter);
+    // app.use('/login', loginRouter);
```
Comment (anishapant21) [inline backend/server.js:27]: .

File: backend/config/database.js (hunk 1 above)
Comment (abroa01) [reply backend/config/database.js:5]: Done

**Review by anishapant21 — APPROVED** [review]

**Pull request merged**